### PR TITLE
fix(redis): go-redis v9 regression missing metrics and reconnect hook (#13415)

### LINKED ISSUE #13415: Missing `argocd_redis_request_total` metric
**Describe the bug**

The metric `argocd_redis_request_total` is not appearing post-upgrade of argocd. 

I've migrated from `2.4.x` to `2.7.0`.

I can see that the metric still exists in the code still, but I no longer see this metric being output.

This is utilizing the official helm chart (`v5.30.1`) which is installing redis locally.

I have verified that the argocd components are talking to the redis instance using `MONITOR` ( https://redis.io/commands/monitor/ ) - i.e. I see output here when I visit the argocd UI.

I have also port-forwarded to the controller-metric svc to confirm that other metrics are available at `/metrics`  (they are, but there's no `redis_*` ones appearing.

**To Reproduce**

Check if the `argocd_redis_request_total` metric increases when communicating with Redis

**Expected behavior**

The `argocd_redis_request_total` metric should increase

### PR DESCRIPTION
**Pull request by phanama**
The update to go-redis v9 broke redis metrics generation. The `BeforeProcess()` and `AfterProcess()` hooks were replaced by `ProcessHook()` and `DialHook()` in v9 ([ref](https://pkg.go.dev/github.com/go-redis/redis/v9#Hook)), which are implemented as empty funcs in our codebase.

This PR moves the implementations to the correct functions and improves tests.

ref: https://github.com/argoproj/argo-cd/commit/7b74e1993e6bd8ed063f4b50414163842e08f359

Fixes https://github.com/argoproj/argo-cd/issues/13415
Most probably also fixes an unreported regression of https://github.com/argoproj/argo-cd/pull/7207

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

**Comment by phanama** [comment]
@crenshaw-dev 
Sorry for mentioning directly, could you please take a look? 🙏 
Redis metrics have been missing since v2.7, I think we need to cherrypick for 2.7-2.9

**Comment by 34fathombelow** [comment]
@leoluz I created a image based on the `phanama:go-redis-v9-regression` branch and tested.  I can confirm that this PR does indeed fix the issue.  When you get a minute can you do a code review?

**Review by 34fathombelow — APPROVED** [review]
LGTM, thanks for the PR great work!

File: util/cache/redis_hook.go [hunk 1]
```diff
@@ -2,13 +2,14 @@ package cache
...
-	"strings"
+	"errors"
+	"net"
 
 	"github.com/redis/go-redis/v9"
 	log "github.com/sirupsen/logrus"
 )
 
-const NoSuchHostErr = "no such host"
+var dnsError *net.DNSError
```
Comment (leoluz) [inline util/cache/redis_hook.go:12]: Is there any good reason to have this declared at the package global? If not, can you move this variable inside the function that depends on it and avoid the global state?

**Review by leoluz — CHANGES_REQUESTED** [review]
Minor comment.. Otherwise LGTM

**Review by leoluz — APPROVED** [review]
LGTM

**Pull request merged**